numpy/core/src/multiarray/refcount.c

```diff
@@ -51,6 +51,31 @@ PyArray_Item_INCREF(char *data, PyArray_Descr *descr)
             PyArray_Item_INCREF(data + offset, new);
         }
     }
+    else if (PyDataType_HASSUBARRAY(descr)) {
+        int size, i, inner_elsize;
+
+        inner_elsize = descr->subarray->base->elsize;
+        /* Subarrays are always contiguous in memory */
+        size = descr->elsize / inner_elsize;
```

Comment:
This divide by zero for `np.dtype(('V0', (1, 2)))`

Reply:
Hmm, I can't work out how to construct that dtype any more

Reply:
Numpy tries really hard to stop me, but I can produce that type with:
```python
dt = np.dtype('V0')
dt = copy.copy(dt)
dt.__setstate__((3, '|', (np.dtype('V0'), (3,)), None, None, 0, 1, 0)))
```

Reply:
Wow, good point. Calculating from shape would be nice, but unfortunately I think the shape is only stored as a python tuple right now, so I doubt that is prettier then even checking explicitly for `inner_elsize == 0`.

I think I would just add it, with a comment that it seems currently not possible. I think the real problem would be a dtype such as `np.dtype([('a', 'i4'), ('b', [('ba', 'f', (5, 0))], (3,3))])` (a subarray containing a 0-size subarray). A simple `V0` cannot hold a referene so it can't reach this far (although that is not obvious).

Reply:
Guess what? Numpy tries really hard to stop you producing that type too, which is clearly a bug.

Reply:
Perhaps an intermediate step of changing to
```
#define PyDataType_ISUNSIZED(dtype) (\
    (dtype)->elsize == 0 && !PyDataType_HASSUBARRAY(dtype) && !PyDataType_HASFIELDS(dtype))
```
Would make these types creatable without going through pickle.

Reply:
Yeah. It is a bit strange that mine was also not creatable (not sure if you trick can manage). For here, I think I would just go with a check for `if elsize == 0: return` since it obviously cannot hold references then, and will be future proof no matter what.

I think "unsized" probably means things like `S0` which actually are used to mean "S<whatever size>"? Would have to look at the code to be quite sure.

Reply:
Hmmm, sorry, I guess your check would likely suffice in most places when it comes to that.

Reply:
To be clear - I'm not saying you should use that check - I'm saying that the existing `PyDataType_ISUNSIZED` is wrong and misfires on empty subarrays and on structures with no fields.

Reply:
> not sure if you trick can manage

Here's the general solution
```
def subarray_dtype(scalar, shape):
        """ like np.dtype((scalar, shape)), but always produces subarrays in corner cases """
	scalar = np.dtype(scalar)
	dt = np.dtype(np.void)
	dt = copy.copy(dt)
	dt.__setstate__((3, '|', (scalar, shape), None, None, scalar.itemsize * np.prod(shape), scalar.alignment, 0))
	return dt
```

Reply:
Moved discussion about the dtype construction pain to #12681, and resolving this comment since you fixed the C code here.